DigitalLearningSolutions.Web/Controllers/RegisterController.cs

```diff
@@ -0,0 +1,164 @@
+            var viewModel = data!.LearnerInformationViewModel;
+            ViewBag.Centres = centresService.GetActiveCentres();
+            ViewBag.JobGroups = jobGroupsService.GetJobGroups();
+
+            return View(viewModel);
+        }
+
+        [ServiceFilter(typeof(RedirectEmptySessionData<DelegateRegistrationData>))]
+        [HttpPost]
+        public IActionResult LearnerInformation(LearnerInformationViewModel model)
+        {
+            if (!ModelState.IsValid)
+            {
+                ViewBag.Centres = centresService.GetActiveCentres();
+                ViewBag.JobGroups = jobGroupsService.GetJobGroups();
+                return View(model);
+            }
+
+            var data = TempData.Get<DelegateRegistrationData>();
+            data!.LearnerInformationViewModel = model;
+            TempData.Set(data);
+
+            return RedirectToAction("Password");
+        }
+
+        [ServiceFilter(typeof(RedirectEmptySessionData<DelegateRegistrationData>))]
+        [HttpGet]
+        public IActionResult Password()
+        {
+            var data = TempData.Get<DelegateRegistrationData>();
+            var viewModel = data!.PasswordViewModel;
+
+            return View(viewModel);
+        }
+
+        [ServiceFilter(typeof(RedirectEmptySessionData<DelegateRegistrationData>))]
+        [HttpPost]
+        public IActionResult Password(PasswordViewModel model)
+        {
+            if (!ModelState.IsValid)
+            {
+                return View(model);
+            }
+            var data = TempData.Get<DelegateRegistrationData>();
+            data!.PasswordViewModel = model;
+            TempData.Set(data);
+
+            return RedirectToAction("Summary");
+        }
+
+        [ServiceFilter(typeof(RedirectEmptySessionData<DelegateRegistrationData>))]
+        [HttpGet]
+        public IActionResult Summary()
+        {
+            var data = TempData.Get<DelegateRegistrationData>();
+            var viewModel = MapToSummary(data!);
+
+            return View(viewModel);
+        }
+
+        [ServiceFilter(typeof(RedirectEmptySessionData<DelegateRegistrationData>))]
+        [HttpPost]
+        public IActionResult Summary(SummaryViewModel model)
+        {
+            if (!ModelState.IsValid)
+            {
+                var data = TempData.Get<DelegateRegistrationData>();
+                var viewModel = MapToSummary(data!);
+                viewModel.Terms = model.Terms;
+                return View(viewModel);
+            }
+
+            // TODO HEEDLS-396 create new user and redirect to confirmation page
+            return View(model);
+        }
+
+        private SummaryViewModel MapToSummary(DelegateRegistrationData data)
+        {
+            var centre = centresService.GetCentreName((int)data.LearnerInformationViewModel.Centre!);
+            var jobGroup = jobGroupsService.GetJobGroupName((int)data.LearnerInformationViewModel.JobGroup!);

```

Comment:
What happens in the UI if these ids don't exist (e.g. cookie has expired)?

Reply:
If the cookie's expired then we will redirect back to the first page of the form before this function gets called. Theoretically though someone could visit the first page to get the cookie and then manually enter the summary URL which would cause a problem at the moment.
I've got 2 ideas for how to deal with that:
* Automatically redirect back to the appropriate page for the first missing or invalid field when loading the summary page
* Display error messages on the summary page for each missing or invalid field
Any preference on which one? I'm leaning towards the first because I don't know if there's an NHS standard way to display errors in tables

Reply:
Yes, I think you're right - the first option sounds good. If it's too time consuming we might be able to leave it, because the probability of this happening should be fairly slim. :)